### DISCUSSION TITLE
(멀티플레이) 텔레포트 동기화

### DISCUSSION THREAD

**Discussion by smilejsu82** [naverz/zepeto-studio-kor#672] [5. Q&A]
월드 1.2.0 입니다 

https://user-images.githubusercontent.com/50188795/165511659-d17a5cac-bfbe-41d1-b269-905c84d4a2b1.mp4

![캡처_2022_04_27_20_34_00_797](https://user-images.githubusercontent.com/50188795/165511667-1d19648d-aad5-41a2-9e5c-2a89e09e67e6.png)

가끔 텔레포트 동기화가 잘 안되는거 같은데 무엇이 잘못되었을까요?

다음과 같은 방법을 사용했습니다 

![image](https://user-images.githubusercontent.com/50188795/165511480-5460c5f0-138d-4250-829e-bf93b810e4c3.png)
![image](https://user-images.githubusercontent.com/50188795/165511499-34bd995e-5cd9-457a-9ab3-3c59366369ef.png)
![image](https://user-images.githubusercontent.com/50188795/165511517-f28fe6c0-f8a2-4dab-af5d-77a3cc6c46fb.png)
![image](https://user-images.githubusercontent.com/50188795/165511552-1231d2a0-0e83-45b2-941b-d48d1fabec94.png)
![image](https://user-images.githubusercontent.com/50188795/165511564-d0039f27-427b-4d4e-b02c-a25437766584.png)
![image](https://user-images.githubusercontent.com/50188795/165511577-15b7156c-6a46-41e0-a472-b29fc353c16a.png)
![image](https://user-images.githubusercontent.com/50188795/165511593-acb6a7bb-144e-4772-80f3-061705716a89.png)
![image](https://user-images.githubusercontent.com/50188795/165511606-a8c3be52-6b8d-4506-a7f0-240642b2ce71.png)
![image](https://user-images.githubusercontent.com/50188795/165511617-e0af3b0b-d230-47f5-9323-72b1281227ab.png)
![image](https://user-images.githubusercontent.com/50188795/165511629-94941211-b4fd-43a4-9f4a-989becf3b57a.png)

다음은 코드 입니다 

공격을 받으면 자신껏만 처리 합니다 이때 체력이 0이 되면 쓰러지는 애니메이션을 합니다 
![image](https://user-images.githubusercontent.com/50188795/165511924-4b7da11f-f88d-44bb-99bc-8756ea023030.png)

그리고 서버로 보내요 
![image](https://user-images.githubusercontent.com/50188795/165511981-00b0a27a-0216-4fae-8a51-0475b918c6b4.png)
![image](https://user-images.githubusercontent.com/50188795/165512007-467963a3-2a91-4b17-b44e-9ab1557ea4cb.png)

서버에서는 그냥 전달만 합니다 
![image](https://user-images.githubusercontent.com/50188795/165512045-efc4baa7-b8dd-4bdb-b5fe-3dabc9d2742e.png)

클라에서는 다시 받아서 내꺼가 아닌것만 피해를 줘요 (리모트 플레이어 적용)
![image](https://user-images.githubusercontent.com/50188795/165512097-ef2687fb-de21-4062-9b9e-792f5b329d24.png)

리모트 플레이어가 피해를 받아 체력이 0이 되면 쓰러지는 애니메이션을 실행 해요 
![image](https://user-images.githubusercontent.com/50188795/165512206-b31444e0-40cf-4fce-940c-667a10f6002a.png)

아까 자신꺼 처리 할때 체력이 0이되면 쓰러지는 애니메이션을 하고 
잠시후 텔레포트 합니다 
![image](https://user-images.githubusercontent.com/50188795/165512371-2c5c3040-b6b5-43c4-88e5-10d0d2889bd2.png)
![image](https://user-images.githubusercontent.com/50188795/165512432-881bd88c-496e-4668-8d61-54c36b028867.png)

그리고 서버로 전송 해요 
![image](https://user-images.githubusercontent.com/50188795/165512478-28115a3b-1c43-4254-a202-bac8862951eb.png)
![image](https://user-images.githubusercontent.com/50188795/165512504-3f8f7fb0-9197-4973-94a0-801db64ca379.png)

서버에서는 다시 전달 만 해요 
![image](https://user-images.githubusercontent.com/50188795/165512558-68711e64-0900-4c6f-ba5a-17e30201d9a7.png)

클라에서는 다시 받아서 리모트 플레이어만 동기화 합니다 
![image](https://user-images.githubusercontent.com/50188795/165512619-a95ba66b-715d-4f92-b8b8-4ff39b981a50.png)

리모트 플레이어도 마찬가지로 텔레포트 해줍니다 
![image](https://user-images.githubusercontent.com/50188795/165512640-f435d3ab-aa61-48c4-9c4a-e432c132aca5.png)

**Comment by minjujuu** [comment]
RemoteTeleport() 함수 로직을 조금 수정하여 while 같은 반복문 안에서 Teleport를 계속 호출하고,
만약 해당 캐릭터가 Teleport 하려던 위치로 이동되었다면 반복문을 멈추는 방식으로 하면 어떨까요?
올려주신 코드에 다른 문제는 없어보이고 간헐적으로 안된다고 하시는 것 같아서
혹시 다른 영향으로 인해 한 번의 Teleport로 캐릭터가 이동되지 않았을 경우에도 재호출을 통해 이동은 되도록 처리해주기 위해 반복문을 이용하면 어떨까 싶습니다!

**Reply by smilejsu82** [reply]
> RemoteTeleport() 함수 로직을 조금 수정하여 while 같은 반복문 안에서 Teleport를 계속 호출하고, 만약 해당 캐릭터가 Teleport 하려던 위치로 이동되었다면 반복문을 멈추는 방식으로 하면 어떨까요? 올려주신 코드에 다른 문제는 없어보이고 간헐적으로 안된다고 하시는 것 같아서 혹시 다른 영향으로 인해 한 번의 Teleport로 캐릭터가 이동되지 않았을 경우에도 재호출을 통해 이동은 되도록 처리해주기 위해 반복문을 이용하면 어떨까 싶습니다!

답변 감사합니다 ~ 
구조를 조금 바꿔 해결 했습니다 ~ 
![캡처_2022_04_28_12_44_18_543](https://user-images.githubusercontent.com/50188795/165672235-c3f16092-f898-48a1-9094-ddc9eacae864.png)

![캡처_2022_04_28_12_44_27_853](https://user-images.githubusercontent.com/50188795/165672242-2d741b70-bfe8-4c9b-9df7-f6844531b805.png)
![캡처_2022_04_28_12_44_34_650](https://user-images.githubusercontent.com/50188795/165672246-48385e37-54e3-4892-a813-dea2a7babbf3.png)
![캡처_2022_04_28_12_44_38_733](https://user-images.githubusercontent.com/50188795/165672249-6914d057-3647-493f-aa4c-b8e085df4998.png)
![캡처_2022_04_28_12_44_42_166](https://user-images.githubusercontent.com/50188795/165672254-f7e99627-6706-4bbe-9529-cbbbb3da07fa.png)
![캡처_2022_04_28_12_44_47_333](https://user-images.githubusercontent.com/50188795/165672258-2503dffd-1cc7-4853-9303-e4546e58e73f.png)
![캡처_2022_04_28_12_44_50_850](https://user-images.githubusercontent.com/50188795/165672260-64e502a2-348f-4a84-9df9-8fd24186843d.png)
![캡처_2022_04_28_12_44_54_116](https://user-images.githubusercontent.com/50188795/165672269-e7be527c-a560-410c-b7fb-a83ace50fdca.png)
![캡처_2022_04_28_12_44_57_983](https://user-images.githubusercontent.com/50188795/165672273-a1ca83f2-1fe8-4e9e-afc0-8c52879144bf.png)
![캡처_2022_04_28_12_45_02_799](https://user-images.githubusercontent.com/50188795/165672277-69f52329-6fdd-432f-9528-29e9739728b5.png)
![캡처_2022_04_28_12_45_06_249](https://user-images.githubusercontent.com/50188795/165672282-401df5e2-c37c-4b8e-81d1-039464e57137.png)
![캡처_2022_04_28_12_45_09_400](https://user-images.githubusercontent.com/50188795/165672288-cee8c389-78f6-4f5a-9924-825bb7a84091.png)
![캡처_2022_04_28_12_45_12_448](https://user-images.githubusercontent.com/50188795/165672297-784d126b-345b-49f4-9ed2-1244810ee9c6.png)
![캡처_2022_04_28_12_45_15_633](https://user-images.githubusercontent.com/50188795/165672304-a2354e63-f435-482d-8f8f-72edd248eaf6.png)
![캡처_2022_04_28_12_45_18_365](https://user-images.githubusercontent.com/50188795/165672313-fbb1ab63-180b-4986-985d-fbbe418ff5df.png)
![캡처_2022_04_28_12_45_21_215](https://user-images.githubusercontent.com/50188795/165672317-dade74a7-350a-45ad-8820-fbf586976ba4.png)
![캡처_2022_04_28_12_45_24_965](https://user-images.githubusercontent.com/50188795/165672325-92c96c27-53b2-41e9-b85c-f920b12bc271.png)
[프레젠테이션1.pdf](https://github.com/naverz/zepeto-studio-kor/files/8579337/1.pdf)

**Reply by smilejsu82** [reply]
https://user-images.githubusercontent.com/50188795/165674099-134cf2cd-0dc9-4a84-8edb-a7038f2e1abf.mp4

https://user-images.githubusercontent.com/50188795/165674115-384c39c1-153e-426e-8e42-48415e267b24.mp4

**Reply by minjujuu** [reply]
해결되어 다행이네요! 상세한 공유 감사합니다~

**Reply by Sonak0930** [reply]
혹시 packet은 어느 라이브러리인지 알 수 있을까요?
